runelite-client/src/main/java/net/runelite/client/plugins/lmsperformancetracker/LmsPerformanceTrackerPlugin.java

```diff
@@ -0,0 +1,336 @@
+		{
+			return;
+		}
+
+		Actor opponent;
+
+		// If the event source is the player, then it is the player interacting with their potential opponent.
+		if (event.getSource() == client.getLocalPlayer())
+		{
+			opponent = event.getTarget();
+		}
+		else if (event.getTarget() == client.getLocalPlayer())
+		{
+			opponent = event.getSource();
+		}
+		else // if neither source or target was the player, skip
+		{
+			return;
+		}
+
+		currentOpponent = (Player) opponent;
+		currentFight = new FightPerformance(client.getLocalPlayer().getName(), currentOpponent.getName());
+		lastFightTime = Instant.now().minusSeconds(NEW_FIGHT_DELAY.getSeconds() - 3);
+	}
+
+	@Subscribe
+	public void onGameTick(GameTick gameTick)
+	{
+		if (isAtLMS())
+		{
+			if (!navButtonShown && config.saveFightHistory())
+			{
+				clientToolbar.addNavigation(navButton);
+				navButtonShown = true;
+			}
+		}
+		else
+		{
+			if (config.restrictToLms())
+			{
+				if (navButtonShown)
+				{
+					clientToolbar.removeNavigation(navButton);
+					navButtonShown = false;
+				}
+				return;
+			}
+		}
+
+		stopFightIfOver();
+
+		if (hasOpponent())
+		{
+			// if there is an opponent, check both players' animations for attacks.
+			// If they attacked, will also add the attack to the current fight stats,
+			// taking into account the opponent's overhead prayer.
+			checkAnimation(true);
+			checkAnimation(false);
+		}
+	}
+
+	// Will set the currentOpponent to null if the player should no longer have a target:
+	// if either player hasn't fought in NEW_FIGHT_DELAY, or either player died.
+	// Will also add the currentFight to fightHistory if the fight ended.
+	private void stopFightIfOver()
+	{
+		// if either player died, end the fight.
+		if (hasOpponent() && currentOpponent.getAnimation() == AnimationID.DEATH)
+		{
+			currentFight.opponentDied();
+			stopCurrentFight();
+		}
+		if (hasOpponent() && client.getLocalPlayer().getAnimation() == AnimationID.DEATH)
+		{
+			currentFight.playerDied();
+			stopCurrentFight();
+		}
+		// If there was no fight actions in the last (NEW_FIGHT_DELAY) seconds, set opponent to
+		// null, which will get set next time the player targets a Player.
+		if (Duration.between(lastFightTime, Instant.now()).compareTo(NEW_FIGHT_DELAY) > 0)
+		{
+			stopCurrentFight();
+		}
+	}
+
+	// Returns true if the player has an opponent.
+	private boolean hasOpponent()
+	{
+		return currentOpponent != null;
+	}
+
+	// check player's animation, and if they are doing an attack animation, add an attack
+	// to the current fight stats. Check opponent's overhead to see if the attack was "successful".
+	// forPlayer bool: when true, checking animations for the local player. False will check the opponent.
+	private void checkAnimation(boolean forPlayer)
+	{
+		Player player = forPlayer ? client.getLocalPlayer() : currentOpponent;
+		Player opponent = forPlayer ? currentOpponent : client.getLocalPlayer();
```

Comment:
I think that these ternaries depending on `forPlayer` make it unnecessarily complex. Maybe you can change some logic so that you can check both at the same time. The are called with `true/false` sequentially at L240 anyway.

Reply:
This is pretty much the only thing remaining aside from the few I disagreed. The Fighter class cleaned up a lot of things nicely. I'm gonna try make the Fighter class implement Player so all of this can really be written together without having to match up instances of Player and Fighter. It might not work as well as I imagine it though so I might end up fixing it another way.